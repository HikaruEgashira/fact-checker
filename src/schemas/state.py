import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel


TABLE_NAME = os.environ.get("TABLE_NAME") or "fact-checker-results"

dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(TABLE_NAME)

FactCheckOutput = Literal["accurate", "inaccurate", "false", "indeterminate"]
Status = Union[Literal["pending"], FactCheckOutput]


class State(BaseModel):
    id: str
    result: Status


def update_state(state: State):
    return table.put_item(Item=state.model_dump())


def get_state(state_id: str):
    response = table.get_item(Key={"id": state_id})
    return State(**response["Item"]) if "Item" in response else None  # type: ignore


# only for testing purposes
def delete_state(state_id: str):
    return table.delete_item(Key={"id": state_id})
