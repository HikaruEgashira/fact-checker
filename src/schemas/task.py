import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel


TABLE_NAME = os.environ.get("TABLE_NAME") or "fact-checker-results"

dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(TABLE_NAME)

FactCheckOutput = Literal["accurate", "inaccurate", "false", "indeterminate"]
Status = Union[Literal["pending"], FactCheckOutput]


class Task(BaseModel):
    task_id: str
    text: str
    result: Status


def update_task(task: Task):
    return table.put_item(Item=task.model_dump())


def get_task(task_id: str):
    response = table.get_item(Key={"task_id": task_id})
    return Task(**response["Item"]) if "Item" in response else None  # type: ignore


# only for testing purposes
def delete_task(task_id: str):
    return table.delete_item(Key={"task_id": task_id})
