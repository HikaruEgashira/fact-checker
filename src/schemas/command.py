import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel, Field

from schemas.state import current_session

QUEUE_NAME = os.environ.get("QUEUE_NAME") or "fact-checker-queue"
REGION = os.environ.get("AWS_REGION") or "ap-northeast-1"

session = boto3.Session()
sqs = session.client("sqs")
sts = session.client("sts")

queue_url = f"https://sqs.{REGION}.amazonaws.com/{sts.get_caller_identity()['Account']}/{QUEUE_NAME}"

# Command pattern
# https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions


class EntryCommand(BaseModel):
    type: Literal["entry"] = "entry"
    prompt: str


class UnknownCommand(BaseModel):
    type: Literal["unknown"] = "unknown"


Command = Union[EntryCommand, UnknownCommand]


class Request(BaseModel):
    command: Command = Field(..., discriminator="type")
    id: str


# Utility functions


def send_command(command: Command):
    request = Request(id=current_session(), command=command)
    return sqs.send_message(QueueUrl=queue_url, MessageBody=request.model_dump_json())


def send_batch(commands: list[Command]):
    return sqs.send_message_batch(
        QueueUrl=queue_url,
        Entries=[
            {
                "Id": str(i),
                "MessageBody": Request(
                    id=current_session(), command=command
                ).model_dump_json(),
            }
            for i, command in enumerate(commands)
        ],
    )
