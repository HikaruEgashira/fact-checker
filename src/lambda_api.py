from pydantic import BaseModel
from aws_lambda_powertools import Logger, Tracer
from aws_lambda_powertools.event_handler import APIGatewayRestResolver, Response
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

from schemas.state import State, update_state, get_state
from schemas.command import FactcheckCommand, send_command

tracer = Tracer()
logger = Logger()
app = APIGatewayRestResolver(enable_validation=True)


class FactCheckRequest(BaseModel):
    prompt: str


class FactCheckState(BaseModel):
    id: str
    status: str
    output: str


@app.post("/fact-check")
@tracer.capture_method
def enqueue_fact_check() -> FactCheckState:
    body = app.current_event.json_body
    prompt = body["prompt"]
    state_id = app.current_event.request_context.request_id

    # Save the state to DynamoDB
    state = State(id=state_id, status="pending", output="")
    update_state(state)

    # Send the state to the queue
    command = FactcheckCommand(prompt=prompt)
    send_command(command)

    return FactCheckState(id=state_id, status="pending", output="")


@app.get("/fact-check/<state_id>")
@tracer.capture_method
def check_state(state_id: str) -> FactCheckState:
    state = get_state(state_id)
    if state:
        return FactCheckState(id=state.id, status=state.status, output=state.output)
    else:
        raise Exception("State not found")


@app.exception_handler(Exception)
def handle_exception(ex: Exception):
    logger.error(ex)
    return Response(status_code=404)


@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_HTTP)
@tracer.capture_lambda_handler
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
