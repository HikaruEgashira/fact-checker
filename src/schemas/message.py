import os
from typing import Literal
from aws_lambda_powertools.utilities.parser import BaseModel
from schemas.aws import session

QUEUE_NAME = os.environ.get("QUEUE_NAME") or "fact-checker-queue"
REGION = os.environ.get("AWS_REGION") or "ap-northeast-1"

sqs = session.client("sqs")
sts = session.client("sts")

queue_url = f"https://sqs.{REGION}.amazonaws.com/{sts.get_caller_identity()['Account']}/{QUEUE_NAME}"


ActionType = Literal["execute"]


class ExecuteMessage(BaseModel):
    type: ActionType = "execute"
    text: str
    task_id: str


def send_message(message: BaseModel):
    message_raw = message.model_dump_json()
    return sqs.send_message(QueueUrl=queue_url, MessageBody=message_raw)
