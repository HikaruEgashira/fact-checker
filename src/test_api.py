import json
from time import sleep
from aws_lambda_powertools.utilities.typing import LambdaContext

from lambda_api import (
    enqueue_fact_check_state,
    check_state_status,
)
from schemas.state import delete_state


context = LambdaContext()
context._aws_request_id = "unique-request-id"


def test_api(snapshot):
    state = "The text to be fact-checked."
    enqueue_response = enqueue_fact_check_state(state, context)
    assert "id" in enqueue_response["body"]
    state_id = json.loads(enqueue_response["body"])["id"]
    check_response = {}
    retry = 10
    while retry > 0:
        check_response = check_state_status(state_id, context)
        if json.loads(check_response["body"])["result"] != "pending":
            break
        sleep(1)
        retry -= 1
    assert check_response["statusCode"] == 200
    assert check_response == snapshot

    delete_state(state_id)
