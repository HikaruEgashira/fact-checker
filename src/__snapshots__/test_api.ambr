# serializer version: 1
# name: test_api
  dict({
    'body': '{"id":"unique-request-id","result":"accurate"}',
    'statusCode': 200,
  })
# ---
